src/v/arch/manifest.cc

```diff
@@ -0,0 +1,206 @@
+}
+
+manifest manifest::difference(const manifest& remote_set) const {
+    if (_ntp != remote_set._ntp && _rev != remote_set._rev) {
+        throw manifest_error("NTPs doesn't match");
+    }
+    manifest result(_ntp, _rev);
+    std::set_difference(
+      begin(),
+      end(),
+      remote_set.begin(),
+      remote_set.end(),
+      result.get_insert_iterator());
+    return result;
+}
+
+ss::future<> manifest::update(ss::input_stream<char>&& is) {
+    iobuf result;
+    auto os = make_iobuf_ref_output_stream(result);
+    co_await ss::copy(is, os);
+    iobuf_istreambuf ibuf(result);
+    std::istream stream(&ibuf);
+    ptree m;
+    boost::property_tree::json_parser::read_json(stream, m);
+    update(m);
+    co_return;
+}
+
+void manifest::update(const manifest::ptree& m) {
+    // ntp from json
+    auto ns = model::ns(m.get<ss::sstring>("namespace"));
+    auto tp = model::topic(m.get<ss::sstring>("topic"));
+    auto pt = model::partition_id(m.get<int32_t>("partition"));
+    _rev = model::revision_id(m.get<int64_t>("revision"));
+    _ntp = model::ntp(ns, tp, pt);
+    segments_map tmp;
+    if (m.count("segments") != 0) {
+        for (const ptree::value_type& it : m.get_child("segments")) {
+            auto coffs = it.second.get<int64_t>("committed_offset");
+            auto boffs = it.second.get<int64_t>("base_offset");
+            segment_meta meta{
+              .is_compacted = it.second.get<bool>("is_compacted"),
+              .size_bytes = it.second.get<size_t>("size_bytes"),
+              .base_offset = model::offset(boffs),
+              .committed_offset = model::offset(coffs),
+              .is_deleted_locally = it.second.get<bool>("deleted"),
+            };
+            tmp.insert(std::make_pair(segment_name(it.first.data()), meta));
+        }
+    }
+    std::swap(tmp, _segments);
+}
+
+std::tuple<ss::input_stream<char>, size_t> manifest::serialize() const {
+    iobuf serialized;
+    iobuf_ostreambuf obuf(serialized);
+    std::ostream os(&obuf);
+    serialize(os, false);
+    size_t size_bytes = serialized.size_bytes();
+    return std::make_tuple(
+      make_iobuf_input_stream(std::move(serialized)), size_bytes);
+}
+
+void manifest::serialize(std::ostream& out, bool pretty) const {
+    ptree m;
+    m.put("namespace", _ntp.ns());
+    m.put("topic", _ntp.tp.topic());
+    m.put("partition", _ntp.tp.partition());
+    m.put("revision", _rev());
+    ptree segments;
+    for (const auto& [sn, meta] : _segments) {
+        ptree it;
+        it.put("is_compacted", meta.is_compacted);
+        it.put("size_bytes", meta.size_bytes);
+        it.put("committed_offset", meta.committed_offset());
+        it.put("base_offset", meta.base_offset());
+        it.put("deleted", meta.is_deleted_locally);
+        segments.push_back(std::make_pair(sn(), std::move(it)));
+    }
+    m.add_child("segments", segments);
+    boost::property_tree::write_json(out, m, pretty);
```

Comment:
interesting!!! how come we don't use `rapidjson` here which is already a dependency.  just wondering.

Reply:
great